xcommon/src/llvm.rs

```diff
@@ -0,0 +1,121 @@
+//! LLVM utilities
+
+use anyhow::{bail, ensure, Context, Result};
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+/// Returns the set of additional libraries that need to be bundled with
+/// the given library, scanned recursively.
+///
+/// Any libraries in `provided_libs_paths` will be treated as available, without
+/// being emitted. Any other library not in `search_paths` or `provided_libs_paths`
+/// will result in an error.
+pub fn list_needed_libs_recursively(
+    lib: &Path,
+    search_paths: &[&Path],
+    provided_libs_paths: &[&Path],
+) -> Result<HashSet<PathBuf>> {
+    // Create a view of all libraries that are available on Android
+    let mut provided = HashSet::new();
+    for path in provided_libs_paths {
+        for lib in find_libs_in_dir(path).with_context(|| {
+            format!("Unable to list available libraries in `{}`", path.display())
+        })? {
+            // libc++_shared is bundled with the NDK but not available on-device
+            if lib != "libc++_shared.so" {
+                provided.insert(lib);
+            }
+        }
+    }
+
+    let mut to_copy = HashSet::new();
+
+    let mut artifacts = vec![lib.to_path_buf()];
+    while let Some(artifact) = artifacts.pop() {
+        for need in list_needed_libs(&artifact).with_context(|| {
+            format!(
+                "Unable to read needed libraries from `{}`",
+                artifact.display()
+            )
+        })? {
+            // c++_shared is available in the NDK but not on-device.
+            // Must be bundled with the apk if used:
+            // https://developer.android.com/ndk/guides/cpp-support#libc
+            let search_paths = if need == "libc++_shared.so" {
+                provided_libs_paths
+            } else {
+                search_paths
+            };
```

Comment:
yep, this needs to go. you can compute the blocklist for android by searching the provided_libs and filtering for `libc++_shared.so`.

Reply:
And then I copypaste that to a list somewhere... And document the manual commands/steps someone needs to go through to regenerate that list? Please no :(

Reply:
Note also that your change implies that we'll add all the Android library paths from `provided_libs_paths` to `search_paths` which pretty much defeat the purpose since there's exactly one library from there that we need to copy, the rest should not be copied but only used to validate that all dynamic-link dependencies are met.

Reply:
provided_libs_paths must already be included in search_paths. the search paths are anything added via `-L`, otherwise there's a bug.

Reply:
They are not, and it should be by design?

Reply:
well, maybe it's just named confusingly. but if the search path is the same that is passed to the linker, then any library that was linked must be in the search path. So this means that `libc++_shared.so` would be in both the search path and the provided path. So the provided path seems like it should be a subset of search path.

Reply:
libc++ is a poor example, since we actually want to remove it from the provided path.

Reply:
I think that's too fragile. Wondering about taking a different approach. How about we get the linker to tell us which libraries it needs from where? It looks like `ld.lld` supports a `--Map=filepath` argument that looks like it would give us just that.